Add `GetDeclaringType` to `PropertyDefinition` and `EventDefinition`.

teo-tsirpanis opened the pull request:
Fixes #60380. Implementation follows the corresponding methods for method and field definitions, except that the binary search is performed on rows of the `PropertyMap` and `EventMap` tables for properties and fields respectively, and we don't account for rows with no properties/events.

teo-tsirpanis committed:
Add `GetDeclaringType` to `PropertyDefinition` and `EventDefinition`.

teo-tsirpanis committed:
Add test.

teo-tsirpanis committed:
Add documentation to all `GetDeclaringType` APIs.

teo-tsirpanis committed:
Fix tests.

We have to return the value of the parent type column in the map table, not the row to the map table itself.

teo-tsirpanis commented:
@dotnet/area-system-reflection-metadata this is ready for review.

jkotas commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Internal/Tables.cs:
```diff
@@ -1343,6 +1384,47 @@ internal int GetPropertyListStartFor(int rowId)
+                return default(TypeDefinitionHandle);
+            }
+
+            int value = this.GetPropertyListStartFor(row);
+            if (value == propertyRowId)
+            {
+                while (row < numOfRows)
+                {
+                    int newRow = row + 1;
+                    value = this.GetPropertyListStartFor(newRow);
+                    if (value == propertyRowId)
+                    {
+                        row = newRow;
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+
```
I do not think this is needed for property map table. Unlike typedef table that this code was copied from, the property map should not have rows for types with no properties.

Dtto for event map

teo-tsirpanis replied:
Done.

steveharter commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeSystem/TypeDefinitionTests.cs:
```diff
@@ -32,5 +32,41 @@ public void ValidateTypeDefinitionIsNestedWindowsProjection()
+            foreach (var typeDefHandle in reader.TypeDefinitions)
+            {
+                var typeDef = reader.GetTypeDefinition(typeDefHandle);
+                foreach (var nestedTypeHandle in typeDef.GetNestedTypes())
```
Can we add a sanity check here to make sure that these foreach statement actually loop through at least one? Like:

```cs
Assert.True(typeDef.GetNestedTypes().Length > 0);
```

teo-tsirpanis replied:
The test assembly actually [dosen't have any nested types](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.Metadata/tests/Resources/Misc/Members.cs). I tried to add some but after recompiling it [this assert](https://github.com/dotnet/runtime/blob/c587256d18fa01834508b0e5128b5192b48c08dc/src/libraries/System.Reflection.Metadata/tests/PortableExecutable/PEReaderTests.cs#L244-L249) failed.

Your suggested assert wouldn't currently work, because not all types have all kinds of members each.

teo-tsirpanis replied:
Added asserts that `GetTypesWithEvents()` and `GetTypesWithProperties()` are not empty, and commented out the asserts on the nested classes.

teo-tsirpanis committed:
Remove unnecessary code.

teo-tsirpanis committed:
Update tests.

We assert that the test assembly has types with events and properties, and acknowledge that it does not have nested types.

jkotas commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeSystem/TypeDefinitionTests.cs:
```diff
@@ -32,5 +33,45 @@ public void ValidateTypeDefinitionIsNestedWindowsProjection()
+            foreach (var typeDefHandle in reader.TypeDefinitions)
+            {
+                var typeDef = reader.GetTypeDefinition(typeDefHandle);
+                // The assembly does not declare any nested types yet.
```
Would it be better to add a nested type to this assembly instead?

teo-tsirpanis replied:
Added a test.

jkotas reviewed (approved):
LGTM

jkotas commented:
@MichalStrehovsky Do you know on top your head where there any places in the AOT compilers that can use these APIs? If yes, we should create an issue to adopt these APIs once they make it into the tools runtime refence assemblies.

teo-tsirpanis committed:
Test `TypeDefinition.GetDeclaringType()`.

jkotas commented on src/libraries/System.Reflection.Metadata/tests/Metadata/TypeSystem/TypeDefinitionTests.cs:
```diff
@@ -32,5 +33,58 @@ public void ValidateTypeDefinitionIsNestedWindowsProjection()
+
+#if NET
+        [ActiveIssue("https://github.com/dotnet/runtime/pull/111642", TestRuntimes.Mono)]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotNativeAot))]
```
All existing tests for System.Reflection.Metadata use standalone assemblies so that they are compatible with NAOT and other form-factors that do not preserve ECMA metadata. We do not want to be introducing the test based on runtime assembly for System.Reflection.Metadata.

Is there a problem with adding some nested types to the existing Members assembly?

teo-tsirpanis replied:
https://github.com/dotnet/runtime/pull/111646#discussion_r1943954036

teo-tsirpanis replied:
> Would it be better to add some nested types to this assembly instead?

That's how I interpreted the "this assembly" part of that comment.

jkotas replied:
I am sorry for not being clear. I meant to add it to the existing Members.dll assembly.

> https://github.com/dotnet/runtime/pull/111646#discussion_r1943954036

Members.dll is old and it was compiled with the .NET Framework compiler. You can use .NET Framework compiler to recompile the .dll to avoid unrelated diffs, e.g.:

C:\runtime\src\libraries\System.Reflection.Metadata\tests\Resources\Misc>c:\Windows\Microsoft.NET\Framework\v4.0.30319\csc /target:library Members.cs

Once you add a nested type to the .dll, there will be a few tests that need updating (e.g. the test that enumerates all members of the .dll), but I do not expect the relocbloc test to fail.

jkotas replied:
I have reproduced it locally. It is fine to update the expected value in the failing test. I have pushed the change into this PR.

teo-tsirpanis replied:
Thank you!

jkotas committed:
Revert "Test `TypeDefinition.GetDeclaringType()`."

This reverts commit 2013a590f935271ac438f1eeb752510d0662e50b.

jkotas committed:
Add test for nested types

jkotas commented:
/ba-g unrelated timeout

jkotas commented:
Thank you!

MichalStrehovsky commented:
> @MichalStrehovsky Do you know on top your head where there any places in the AOT compilers that can use these APIs? If yes, we should create an issue to adopt these APIs once they make it into the tools runtime refence assemblies.

Nothing comes to mind. We ran into lack of this API in the ILTrim hackathon.
